### PR TITLE
Refactor PackageLoader, fix NodeModel libraries not being loaded from packages

### PR DESCRIPTION
**Pull request by pboyer**
### Issue

During the recent core modularization refactor, the `NodeModel` loading logic was modified but the `PackageLoader` was not updated.  As a result, as @manuKrakow has pointed out, `NodeModel` libraries were no longer being loaded (MAGN-6721).

### Fix

Beyond fixing this issue, I've greatly cleaned up the dependencies for `PackageLoader.`  Now, `PackageLoader` only requires a few mockable interfaces in order to run `LoadAll`.  

All the "real work" of loading a `Package`, like loading the custom nodes and node model libraries is moved into `DynamoModel.`  Instead of injecting dependencies, in the `Package.Load` call, one simply uses the `PackageLoader.Load(Package)` API.  

One could spend more time cleaning and refactoring this work, but this greatly cleans it up while fixing a critical issue.

### Manual testing

I tested this extensively in Revit 2015 and standalone with no issues.

### Reviewer

- [ ] @ikeough

### CODE REVIEW

File: src/DynamoCore/PackageManager/Package.cs [hunk 1]
```diff
@@ -241,51 +217,6 @@ public IEnumerable<string> EnumerateAssemblyFilesInBinDirectory()
...
-                {
-                    var assemblyNodes = new List<TypeLoadData>();
-                    loader.LoadNodesFromAssembly(assem, context, assemblyNodes, new List<TypeLoadData>());
-                    return assemblyNodes;
-                });
-
-            foreach (var node in nodes)
-                LoadedTypes.Add(node.Type);
-        }
-
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/Package.cs]: All of this logic is moved into `DynamoModel`.

File: src/DynamoCore/PackageManager/Package.cs [hunk 2]
```diff
@@ -308,10 +239,10 @@ internal void AddAssemblies(IEnumerable<PackageAssembly> assems)
         }
 
         /// <summary>
-        /// Loads all possible assemblies in node library and returns the list of loaded node library assemblies
+        ///     Enumerates all assemblies in the package
         /// </summary>
         /// <returns>The list of all node library assemblies</returns>
-        private IEnumerable<PackageAssembly> LoadAssembliesInBinDirectory()
+        internal IEnumerable<PackageAssembly> EnumerateAssembliesInBinDirectory()
         {
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/Package.cs]: This is now called from `PackageLoader`, so it is now internal.

File: src/DynamoCore/PackageManager/PackageLoader.cs [hunk 3]
```diff
@@ -14,17 +16,15 @@ namespace Dynamo.PackageManager
...
-        public LibraryServices LibraryServices { get; set; }
-        public DynamoLoader Loader { get; set; }
-        public CustomNodeManager CustomNodeManager { get; set; }
     }
 
     public class PackageLoader : LogSourceBase
     {
+        private readonly ObservableCollection<Package> localPackages = new ObservableCollection<Package>();
+        public ObservableCollection<Package> LocalPackages { get { return localPackages; } }
+
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageLoader.cs]: I just moved this to the top of the body of the class definition.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 4]
```diff
@@ -261,24 +261,9 @@ internal void DownloadAndInstall(PackageDownloadHandle packageDownloadHandle)
...
-                                    Loader = loader,
-                                    LibraryServices = libraryServices,
-                                    Context = DynamoViewModel.Model.Context,
-                                    IsTestMode = DynamoModel.IsTestMode,
-                                    CustomNodeManager = DynamoViewModel.Model.CustomNodeManager
-                                };
-
-                                downloadPkg.LoadIntoDynamo(loadPackageParams, logger);
-
-                                DynamoViewModel.Model.PackageLoader.LocalPackages.Add(downloadPkg);
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: A good example of how much simpler the API for loading a package is now. :)

File: test/DynamoCoreTests/PackageManager/PackageLoaderTests.cs [hunk 5]
```diff
@@ -116,7 +93,6 @@ public void GetOwnerPackageReturnsPackageForValidFunctionDefinition()
                 Assert.AreEqual(pkg.Name, foundPkg.Name);
                 Assert.IsTrue(pkg.Name == foundPkg.Name);
             }
-        
         }
```
Comment (pboyer) [inline test/DynamoCoreTests/PackageManager/PackageLoaderTests.cs]: All of the tests run successfully.

File: src/DynamoCore/PackageManager/PackageLoader.cs [hunk 6]
```diff
@@ -14,17 +16,15 @@ namespace Dynamo.PackageManager
...
     public struct LoadPackageParams
     {
-        public string Context { get; set; }
-        public bool IsTestMode { get; set; }
         public IPreferences Preferences { get; set; }
         public IPathManager PathManager { get; set; }
-        public LibraryServices LibraryServices { get; set; }
-        public DynamoLoader Loader { get; set; }
-        public CustomNodeManager CustomNodeManager { get; set; }
     }
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageLoader.cs]: I'm leaving this class for now as it leaves things easy to refactor.

File: src/DynamoCore/PackageManager/PackageLoader.cs [hunk 7]
```diff
@@ -34,13 +39,107 @@ public PackageLoader(string overridePackageDirectory)
                 Directory.CreateDirectory(RootPackagesDirectory);
         }
 
-        private readonly ObservableCollection<Package> localPackages = new ObservableCollection<Package>();
-        public ObservableCollection<Package> LocalPackages { get { return localPackages; } }
+        private void OnPackageAdded(Package pkg)
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageLoader.cs]: Let's remove `ObservableCollection` here and prevent collection mutation from outside the class.

File: src/DynamoCoreWpf/ViewModels/PackageManager/InstalledPackagesViewModel.cs [hunk 8]
```diff
@@ -36,29 +36,17 @@ private void InitializeLocalPackages()
...
+                LocalPackages.Add(new PackageViewModel(dynamoViewModel, pkg));
+                RaisePropertyChanged("LocalPackages");
+            };
 
-            RaisePropertyChanged("LocalPackages");
+            this.Model.PackageRemoved += (pkg) =>
+            {
+                LocalPackages.Remove(_localPackages.ToList().First(x => x.Model == pkg));
+                RaisePropertyChanged("LocalPackages");
+            };
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/PackageManager/InstalledPackagesViewModel.cs]: We've removed the `ObservableCollection` from `PackageLoader`, so this needs to be cleaned up.

File: test/DynamoCoreTests/CustomNodeWorkspaceOpening.cs [hunk 9]
```diff
@@ -55,15 +55,14 @@ public void CanOpenCustomNodeWorkspace()
...
         {
             var examplePath = Path.Combine(GetTestDirectory(), @"core\combine", "Sequence2.dyf");
             ViewModel.OpenCommand.Execute(examplePath);
             
-            ViewModel.SearchViewModel.SearchAndUpdateResults("Sequence2");
-            Assert.AreEqual(1, ViewModel.SearchViewModel.SearchResults.Count);
-            Assert.AreEqual("Sequence2", ViewModel.SearchViewModel.SearchResults[0].Model.Name);
+            var res = ViewModel.Model.SearchModel.Search("Sequence2");
+            Assert.AreEqual(1, res.Count());
+            Assert.AreEqual("Sequence2", res.First().Name);
```
Comment (pboyer) [inline test/DynamoCoreTests/CustomNodeWorkspaceOpening.cs]: I couldn't resist.  Fixed this test.

File: src/DynamoCore/PackageManager/PackageLoader.cs [hunk 10]
```diff
@@ -34,13 +39,107 @@ public PackageLoader(string overridePackageDirectory)
                 Directory.CreateDirectory(RootPackagesDirectory);
         }
 
-        private readonly ObservableCollection<Package> localPackages = new ObservableCollection<Package>();
-        public ObservableCollection<Package> LocalPackages { get { return localPackages; } }
```
Comment (ikeough) [inline src/DynamoCore/PackageManager/PackageLoader.cs]: I assume that nothing was observing this collection?

File: src/DynamoCoreWpf/ViewModels/PackageManager/InstalledPackagesViewModel.cs (hunk 8 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/PackageManager/InstalledPackagesViewModel.cs]: Ah yes. You were observing it.

**Comment by ikeough** [comment]
@pboyer much cleaner. Thanks. Just a couple of small questions, then LGTM.

**Comment by jnealb** [comment]
@pboyer @ikeough Folded back into 8.0? Tests still run ok?

File: src/DynamoCore/PackageManager/PackageLoader.cs (hunk 6 above)
Comment (Benglin) [inline src/DynamoCore/PackageManager/PackageLoader.cs]: Ouch, my class! :p

**Pull request merged**